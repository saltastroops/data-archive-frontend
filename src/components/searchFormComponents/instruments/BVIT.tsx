import * as React from "react";
import { BVITFilter, IBVIT } from "../../../utils/ObservationQueryParameters";
import { MainGrid, SubGrid } from "../../basicComponents/Grids";
import SelectField, { AnyOption } from "../../basicComponents/SelectField";

interface IBvitProps {
  bvit: IBVIT;
  onChange: (value: any) => void;
}

/**
 * A form for selecting BVIT-related search parameters.
 */
const BVIT = (props: IBvitProps) => {
  const { onChange, bvit } = props;

  // Function for handling change events
  const change = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const name = e.target.name;
    const value = e.target.value;
    onChange({
      ...bvit,
      [name]: value
    });
  };

  const modes = ["Imaging", "Streaming"];
  const filters: BVITFilter[] = ["B", "H-alpha", "Open", "R", "U", "V"];

  return (
    <MainGrid>
      {/* Mode */}
      <SubGrid>
        <p>Mode</p>
        <SelectField name={"mode"} value={bvit.mode} onChange={change}>
          <AnyOption />
          {modes.map(mode => (
            <option key={mode} value={mode}>
              {mode}
            </option>
          ))}
        </SelectField>
      </SubGrid>

      {/* Filter */}
      <SubGrid>
        <p>Filter</p>
        <SelectField name={"filter"} value={bvit.filter} onChange={change}>
          <AnyOption />
          {filters.map(filter => (
            <option key={filter} value={filter}>
              {filter}
            </option>
          ))}
        </SelectField>
      </SubGrid>
    </MainGrid>
  );
};

export default BVIT;
