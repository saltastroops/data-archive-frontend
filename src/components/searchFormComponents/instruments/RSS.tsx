import * as React from "react";
import { ChangeEvent } from "react";
import styled from "styled-components";
import {
  IRSS,
  IRSSModes,
  RSSFabryPerotMode,
  RSSGrating,
  RSSInstrumentMode,
  RSSPolarimetryMode
} from "../../../utils/ObservationQueryParameters";
import { MainGrid, SubGrid } from "../../basicComponents/Grids";
import { LargeCheckbox } from "../../basicComponents/LargeCheckbox";
import SelectField, { AnyOption } from "../../basicComponents/SelectField";
import PolarimetryModesSelector from "../PolarimetryModesSelector";

interface IRssProps {
  rss: IRSS;
  onChange: (value: any) => void;
}

type OnChangeFunction = (value: string) => void;

const ModePropertiesDiv = styled.div`
  p:not(:first-of-type) {
    margin-top: 10px;
  }
`;

/**
 * A form selecting RSS-related search parameters.
 */
const RSS = (props: IRssProps) => {
  const { onChange, rss } = props;
  const modes: IRSSModes = rss.modes || {
    errors: {},
    names: new Set<RSSInstrumentMode>()
  };
  const modeNames = modes.names || new Set<RSSInstrumentMode>();

  const change = (name: string, value: any) => {
    onChange({
      ...rss,
      [name]: value
    });
  };

  /**
   * Handle an instrument change. All properties related to the (previous)
   * instrument are deleted.
   */
  const onModeChange = (e: ChangeEvent<HTMLInputElement>) => {
    // Update the selected mode names
    const updatedModeNames = new Set<RSSInstrumentMode>(modes.names);
    if (e.target.checked) {
      updatedModeNames.add(e.target.name as RSSInstrumentMode);
    } else {
      updatedModeNames.delete(e.target.name as RSSInstrumentMode);
    }

    // Update the modes
    const updatedModes = {
      ...modes,
      names: updatedModeNames
    };
    if (!hasFabryPerotMode) {
      delete updatedModes.fabryPerotMode;
    }
    if (!hasGrating) {
      delete updatedModes.grating;
    }
    if (!hasPolarimetryModes) {
      delete updatedModes.polarimetryModes;
    }

    onChange({
      errors: {
        ...rss.errors
      },
      modes: updatedModes
    });
  };

  const hasFabryPerotMode =
    modeNames.has("Fabry Perot") || modeNames.has("FP polarimetry");

  const hasGrating =
    modeNames.has("Spectroscopy") ||
    modeNames.has("Spectropolarimetry") ||
    modeNames.has("MOS") ||
    modeNames.has("MOS polarimetry");

  const hasPolarimetryModes =
    modeNames.has("Polarimetric imaging") ||
    modeNames.has("Spectropolarimetry") ||
    modeNames.has("MOS polarimetry") ||
    modeNames.has("FP polarimetry");

  const modesPropertyChange = (property: string, value: any) => {
    onChange({
      ...rss,
      modes: {
        ...rss.modes,
        [property]: value
      }
    });
  };

  const instrumentModes: RSSInstrumentMode[] = [
    "Fabry Perot",
    "FP polarimetry",
    "Imaging",
    "MOS",
    "MOS polarimetry",
    "Polarimetric imaging",
    "Spectropolarimetry",
    "Spectroscopy"
  ];

  const detectorModes = [
    "Normal",
    "Frame Transfer",
    "Slot Mode",
    "Shuffle",
    "Drift Scan"
  ];

  return (
    <>
      <MainGrid>
        <SubGrid>
          <p>Mode</p>
          {instrumentModes.map(m => (
            <div key={m}>
              <label>
                <LargeCheckbox
                  name={m}
                  type="checkbox"
                  checked={modeNames.has(m)}
                  onChange={onModeChange}
                />
                {m}
              </label>
            </div>
          ))}
        </SubGrid>
        <SubGrid>
          <ModePropertiesDiv>
            {/* Detector mode */}
            <p>Detector Mode</p>
            <SelectField
              name={"detectorMode"}
              value={rss.detectorMode}
              onChange={e => change("detectorMode", e.target.value)}
            >
              <AnyOption />
              {detectorModes.map(detectorMode => (
                <option key={detectorMode} value={detectorMode}>
                  {detectorMode}
                </option>
              ))}
            </SelectField>

            {/* Grating */}
            {hasGrating && (
              <>
                <p>Grating</p>
                <GratingSelect
                  grating={modes.grating}
                  onChange={(value: string) =>
                    modesPropertyChange("grating", value)
                  }
                />
              </>
            )}

            {/* Fabry-Perot mode */}
            {hasFabryPerotMode && (
              <>
                <p>Fabry-Perot mode</p>
                <FabryPerotModeSelect
                  fabryPerotMode={modes.fabryPerotMode}
                  onChange={(value: string) =>
                    modesPropertyChange("fabryPerotMode", value)
                  }
                />
              </>
            )}

            {/* Polarimetry mode */}
            {hasPolarimetryModes && (
              <>
                <p>Polarimetry mode</p>
                <PolarimetryModesSelector
                  polarimetryModes={modes.polarimetryModes}
                  onChange={(value: Set<RSSPolarimetryMode>) =>
                    modesPropertyChange("polarimetryModes", value)
                  }
                />
              </>
            )}
          </ModePropertiesDiv>
        </SubGrid>
      </MainGrid>
    </>
  );
};

const FabryPerotModeSelect = ({
  fabryPerotMode,
  onChange
}: {
  fabryPerotMode?: RSSFabryPerotMode;
  onChange: OnChangeFunction;
}) => {
  const handleChangeEvent = (e: ChangeEvent<HTMLSelectElement>) =>
    onChange(e.target.value);

  return (
    <SelectField onChange={handleChangeEvent} value={fabryPerotMode || ""}>
      <AnyOption />
      <option value="LR">Low resolution</option>
      <option value="MR">Medium resolution</option>
      <option value="HR">High resolution</option>
      <option value="TF">Tunable filter</option>
    </SelectField>
  );
};

const GratingSelect = ({
  grating,
  onChange
}: {
  grating?: RSSGrating;
  onChange: OnChangeFunction;
}) => {
  const handleChangeEvent = (e: ChangeEvent<HTMLSelectElement>) =>
    onChange(e.target.value);

  return (
    <SelectField onChange={handleChangeEvent} value={grating || ""}>
      <AnyOption />
      <option value="Open">Open</option>
      <option value="pg0300">PG0300</option>
      <option value="pg0900">PG0900</option>
      <option value="pg1300">PG1300</option>
      <option value="pg1800">PG1800</option>
      <option value="pg2300">PG2300</option>
      <option value="ph3000">PG3000</option>
    </SelectField>
  );
};

export default RSS;
