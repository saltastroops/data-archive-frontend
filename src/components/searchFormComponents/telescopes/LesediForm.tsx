import * as React from "react";
import {
  IHIPPO,
  IInstrument,
  ILesedi,
  InstrumentName,
  ISHOC,
  ISpUpNIC
} from "../../../utils/ObservationQueryParameters";
import { MainGrid, SubGrid } from "../../basicComponents/Grids";
import SelectField, { AnyOption } from "../../basicComponents/SelectField";
import HIPPO from "../instruments/HIPPO";
import SHOC from "../instruments/SHOC";
import SpUpNIC from "../instruments/SpUpNIC";

/**
 * Return the form for a given instrument.
 *
 * Parameters:
 * -----------
 * instrument:
 *     The instrument.
 *
 * Returns:
 * --------
 * The form component.
 */
export const lesediInstrumentsSwitcher = (
  instrument: IInstrument,
  onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void
) => {
  const name = instrument && instrument.name;
  switch (name) {
    case "SpUpNIC": {
      return <SpUpNIC spUpNic={instrument as ISpUpNIC} onChange={onChange} />;
    }
    case "HIPPO": {
      return <HIPPO hippo={instrument as IHIPPO} onChange={onChange} />;
    }
    case "SHOC": {
      return <SHOC shoc={instrument as ISHOC} onChange={onChange} />;
    }
    default:
      return null;
  }
};

interface ILesediFormProps {
  lesedi: ILesedi;
  onChange: (key: string, value: any) => void;
}

/**
 * A form for selecting Lesedi-related search parameters.
 */
const LesediForm = (props: ILesediFormProps) => {
  const { lesedi, onChange } = props;

  // Function for changing the instrument
  const changeInstrument = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const value = e.target.value;
    onChange("instrument", { name: value });
  };

  // Function for changing the instrument mode
  const changeMode = (value: any) => {
    onChange("instrument", {
      ...lesedi.instrument,
      ...value
    });
  };

  const instruments: InstrumentName[] = ["HIPPO", "SHOC", "SpUpNIC"];

  return (
    <>
      <MainGrid>
        <SubGrid>
          <p>Instrument</p>
          <SelectField name={"instrument"} onChange={changeInstrument}>
            <AnyOption />
            {instruments.map(instrument => (
              <option key={instrument} value={instrument}>
                {instrument}
              </option>
            ))}
          </SelectField>
        </SubGrid>
      </MainGrid>
      {lesediInstrumentsSwitcher(lesedi.instrument, changeMode)}
    </>
  );
};

export default LesediForm;
