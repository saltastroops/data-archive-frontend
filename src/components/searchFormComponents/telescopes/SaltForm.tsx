import * as React from "react";
import {
  IBVIT,
  IHRS,
  IInstrument,
  InstrumentName,
  IRSS,
  ISALT,
  ISalticam
} from "../../../utils/ObservationQueryParameters";
import { MainGrid, SubGrid } from "../../basicComponents/Grids";
import SelectField, { AnyOption } from "../../basicComponents/SelectField";
import BVIT from "../instruments/BVIT";
import HRS from "../instruments/HRS";
import RSS from "../instruments/RSS";
import Salticam from "../instruments/Salticam";

/**
 * Return the form for a given instrument.
 *
 * Parameters:
 * -----------
 * instrument:
 *     The instrument.
 *
 * Returns:
 * --------
 * The form component.
 */
export const saltInstrumentsSwitcher = (
  instrument: IInstrument,
  onChange: (value: any) => void
) => {
  const name = instrument && instrument.name;
  switch (name) {
    case "RSS": {
      return <RSS rss={instrument as IRSS} onChange={onChange} />;
    }
    case "HRS": {
      return <HRS hrs={instrument as IHRS} onChange={onChange} />;
    }
    case "BVIT": {
      return <BVIT bvit={instrument as IBVIT} onChange={onChange} />;
    }
    case "Salticam": {
      return (
        <Salticam salticam={instrument as ISalticam} onChange={onChange} />
      );
    }
    default:
      return null;
  }
};

interface ISaltFormProps {
  salt: ISALT;
  onChange: (key: string, value: any) => void;
}

/**
 * A form for selecting SALT-related search parameters.
 */
const SaltForm = (props: ISaltFormProps) => {
  const { salt, onChange } = props;
  const instrument = salt.instrument || { name: "" };

  // Function for handling instrument selection
  const changeInstrument = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const value = e.target.value;
    onChange("instrument", { name: value });
  };

  // Function for handling changes to instrument-related search parameters
  const changeInstrumentParameters = (value: any) => {
    onChange("instrument", {
      ...salt.instrument,
      ...value
    });
  };

  const instruments: InstrumentName[] = ["HRS", "RSS", "BVIT", "Salticam"];

  return (
    <>
      <MainGrid>
        <SubGrid>
          <p>Instrument</p>
          <SelectField
            name={"instrument"}
            onChange={changeInstrument}
            value={instrument.name || ""}
          >
            <AnyOption />
            {instruments.map(instr => (
              <option key={instr} value={instr}>
                {instr}
              </option>
            ))}
          </SelectField>
        </SubGrid>
      </MainGrid>
      {saltInstrumentsSwitcher(salt.instrument, changeInstrumentParameters)}
    </>
  );
};

export default SaltForm;
