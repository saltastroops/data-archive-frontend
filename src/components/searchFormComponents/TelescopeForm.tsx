import * as React from "react";
import {
  ILesedi,
  IOneDotNineM,
  ISALT,
  ITelescope,
  TelescopeName
} from "../../utils/ObservationQueryParameters";
import { MainGrid, SubGrid } from "../basicComponents/Grids";
import SelectField, { AnyOption } from "../basicComponents/SelectField";
import LesediForm from "./telescopes/LesediForm";
import OneNineMForm from "./telescopes/OneNineMForm";
import SaltForm from "./telescopes/SaltForm";

const TELESCOPES: TelescopeName[] = ["SALT", "1.9 m", "Lesedi"];

interface ITelescopeFormProps {
  telescope?: ITelescope;
  onChange: (value: any) => void;
}

/**
 * A form for selecting telescope-related search parameters.
 */
class TelescopeForm extends React.Component<ITelescopeFormProps, {}> {
  render() {
    const { telescope, onChange } = this.props;

    // Function for updating telescope-related parameters
    const changeTelescope = (e: React.ChangeEvent<HTMLSelectElement>) => {
      const value = e.target.value;
      onChange({
        name: value
      });
    };

    // Function for updating instrument-related properties
    const changeTelescopeParameter = (key: string, value: any) => {
      onChange({
        ...telescope,
        [key]: value
      });
    };

    const name = (telescope && telescope.name) || "";
    return (
      <>
        <MainGrid>
          <SubGrid>
            <p>Telescope</p>
            <SelectField
              name={"telescope"}
              onChange={changeTelescope}
              value={name || ""}
            >
              <AnyOption />
              {TELESCOPES.map(t => (
                <option key={t} value={t}>
                  {t}
                </option>
              ))}
            </SelectField>
          </SubGrid>
        </MainGrid>
        {name === "SALT" && (
          <SaltForm
            salt={telescope as ISALT}
            onChange={changeTelescopeParameter}
          />
        )}
        {name === "Lesedi" && (
          <LesediForm
            lesedi={telescope as ILesedi}
            onChange={changeTelescopeParameter}
          />
        )}
        {name === "1.9 m" && (
          <OneNineMForm
            oneNineM={telescope as IOneDotNineM}
            onChange={changeTelescopeParameter}
          />
        )}
      </>
    );
  }
}

/**
 * Validate the given telescope-related search parameters and, if need be, add
 * error messages to them.
 */
export const validatedTelescope = (telescope?: ITelescope) => {
  if (telescope) {
    return {
      ...telescope,
      errors: {}
    };
  } else {
    return telescope;
  }
};

export default TelescopeForm;
