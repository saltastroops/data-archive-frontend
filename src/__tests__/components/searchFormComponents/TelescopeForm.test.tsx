import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import BVIT from "../../../components/searchFormComponents/instruments/BVIT";
import RSS from "../../../components/searchFormComponents/instruments/RSS";
import SHOC from "../../../components/searchFormComponents/instruments/SHOC";
import TelescopeForm from "../../../components/searchFormComponents/TelescopeForm";
import LesediForm from "../../../components/searchFormComponents/telescopes/LesediForm";
import OneNineMForm from "../../../components/searchFormComponents/telescopes/OneNineMForm";
import SaltForm from "../../../components/searchFormComponents/telescopes/SaltForm";

const onChange = jest.fn();

describe("TelescopeForm", () => {
  const wrapper = mount(<TelescopeForm onChange={onChange} telescope={{}} />);

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should have a select element with name 'telescope'", () => {
    const telescope = wrapper.find('select[name="telescope"]');
    expect(telescope).toBeDefined();
  });

  it("should have only one child if no telescope is selected", () => {
    const telescope = wrapper.find("div.select").children();
    expect(telescope.length).toEqual(1);
  });

  it("should have two children if a telescope is selected", () => {
    const teleWrapper = mount(
      <TelescopeForm onChange={onChange} telescope={{ name: "SALT" }} />
    );
    const telescope = teleWrapper.find("div.select").children();
    expect(telescope.length).toEqual(2);
  });
});

describe("Telescope form rendering instrument", () => {
  const wrapper = mount(
    <TelescopeForm onChange={onChange} telescope={{ name: "SALT" }} />
  );
  it("should render", () => {
    expect(wrapper).toBeDefined();
  });
  it("should have two children if telescope is selected", () => {
    const telescope = wrapper.find("div.select").children();
    expect(telescope.length).toEqual(2);
  });
  it("should have second child with name instrument ", () => {
    const telescope = wrapper
      .find("div.select")
      .children()
      .get(1);
    expect(telescope).toBeDefined();
    expect(telescope.props.name).toEqual("instrument");
  });

  it("should render have SALT Form if salt is selected", () => {
    const salt = shallow(
      <TelescopeForm onChange={onChange} telescope={{ name: "SALT" }} />
    );
    expect(salt).toBeDefined();
    expect(salt.find(SaltForm).exists()).toBeTruthy();

    // only SaltForm is rendered
    expect(salt.find(OneNineMForm).exists()).toBeFalsy();
    expect(salt.find(LesediForm).exists()).toBeFalsy();
  });
  it("should render have respective form for respective telescope", () => {
    const salt = shallow(
      <TelescopeForm onChange={onChange} telescope={{ name: "SALT" }} />
    );
    const lesedi = shallow(
      <TelescopeForm onChange={onChange} telescope={{ name: "Lesedi" }} />
    );
    const oneNine = shallow(
      <TelescopeForm onChange={onChange} telescope={{ name: "1.9 m" }} />
    );

    expect(salt).toBeDefined();
    expect(lesedi).toBeDefined();
    expect(oneNine).toBeDefined();

    expect(salt.find(SaltForm).exists()).toBeTruthy();
    expect(lesedi.find(LesediForm).exists()).toBeTruthy();
    expect(oneNine.find(OneNineMForm).exists()).toBeTruthy();

    // Should not be rendered
    expect(salt.find(OneNineMForm).exists()).toBeFalsy();
    expect(lesedi.find(OneNineMForm).exists()).toBeFalsy();
    expect(salt.find(LesediForm).exists()).toBeFalsy();
    expect(oneNine.find(LesediForm).exists()).toBeFalsy();
    expect(lesedi.find(SaltForm).exists()).toBeFalsy();
    expect(oneNine.find(SaltForm).exists()).toBeFalsy();
  });
});

describe("Telescope form rendering", () => {
  const wrapper = mount(
    <TelescopeForm
      onChange={onChange}
      telescope={{ name: "SALT", instrument: { name: "RSS" } }}
    />
  );

  it("should render only RSS if RSS is selected", () => {
    const rss = mount(
      <TelescopeForm
        onChange={onChange}
        telescope={{ name: "SALT", instrument: { name: "RSS" } }}
      />
    );
    expect(rss).toBeDefined();
    expect(rss.find(RSS).exists()).toBeTruthy();

    // only RSS is rendered
    expect(rss.find(SHOC).exists()).toBeFalsy();
    expect(rss.find(BVIT).exists()).toBeFalsy();
  });

  it("should render correctly", () => {
    expect(
      toJson(
        shallow(
          <TelescopeForm
            onChange={onChange}
            telescope={{ name: "SALT", instrument: { name: "RSS" } }}
          />
        )
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <TelescopeForm
            onChange={onChange}
            telescope={{ name: "Lesedi", instrument: { name: "SHOC" } }}
          />
        )
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <TelescopeForm
            onChange={onChange}
            telescope={{ name: "SALT", instrument: { name: "HRS" } }}
          />
        )
      )
    ).toMatchSnapshot();

    expect(
      toJson(
        shallow(
          <TelescopeForm
            onChange={onChange}
            telescope={{
              instrument: { name: "RSS", mode: "Fabry Perot" },
              name: "SALT"
            }}
          />
        )
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <TelescopeForm
            onChange={onChange}
            telescope={{
              instrument: {
                detectorMode: "Normal",
                mode: "Fabry Perot",
                name: "RSS"
              },
              name: "SALT"
            }}
          />
        )
      )
    ).toMatchSnapshot();
  });
});

// TODO: test state update correctly
describe("Telescope state can update ", () => {
  it("should call onChange method with correct arguments", () => {
    const wrapper = mount(<TelescopeForm onChange={onChange} telescope={{}} />);
    let value: any;
    let event: any;
    const telescopeSelect = wrapper.find("div.select");
    const telescope = telescopeSelect.find("select");
    value = "SALT";
    event = { target: { value, name: "telescope" } };
    telescope.simulate("change", event);
    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith({ name: "SALT" });

    value = "Lesedi";
    event = { target: { value, name: "telescope" } };
    telescope.simulate("change", event);
    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith({ name: "Lesedi" });
  });
});
