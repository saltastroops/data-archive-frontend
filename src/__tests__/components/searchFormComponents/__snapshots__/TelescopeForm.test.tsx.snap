// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Telescope form rendering should render correctly 1`] = `
<Fragment>
  <styled.div>
    <TelescopeSelector
      onSelect={[Function]}
      telescopes={
        Array [
          "SALT",
        ]
      }
    />
    <InstrumentSelector
      instruments={
        Array [
          "RSS",
        ]
      }
      onSelect={[Function]}
      selectedTelescopes={
        Array [
          "SALT",
        ]
      }
    />
    <InstrumentModeSelector
      instrumentModes={
        Array [
          "MOS",
          "Fabry Perot",
          "Spectropolarimetry",
        ]
      }
      onSelect={[Function]}
      selectedInstruments={
        Array [
          "RSS",
        ]
      }
      selectedTelescopes={
        Array [
          "SALT",
        ]
      }
    />
    <DetectorModeSelector
      detectorModes={
        Array [
          "All",
        ]
      }
      onSelect={[Function]}
      selectedInstruments={
        Array [
          "RSS",
        ]
      }
      selectedTelescopes={
        Array [
          "SALT",
        ]
      }
    />
    <Filters
      filters={
        Array [
          "All",
        ]
      }
      instruments={
        Array [
          "RSS",
        ]
      }
      onSelect={[Function]}
    />
    <RssFabryPerotModeSelector
      onSelect={[Function]}
    />
    <RssGratingSelector
      onSelect={[Function]}
    />
    <RssPolarimetryModeSelector
      onSelect={[Function]}
    />
  </styled.div>
</Fragment>
`;

exports[`Telescope form rendering should render correctly 2`] = `
<Fragment>
  <styled.div>
    <TelescopeSelector
      onSelect={[Function]}
      telescopes={
        Array [
          "Lesedi",
        ]
      }
    />
    <InstrumentSelector
      instruments={
        Array [
          "SHOC",
        ]
      }
      onSelect={[Function]}
      selectedTelescopes={
        Array [
          "Lesedi",
        ]
      }
    />
    <InstrumentModeSelector
      instrumentModes={
        Array [
          "All",
        ]
      }
      onSelect={[Function]}
      selectedInstruments={
        Array [
          "SHOC",
        ]
      }
      selectedTelescopes={
        Array [
          "Lesedi",
        ]
      }
    />
    <DetectorModeSelector
      detectorModes={
        Array [
          "All",
        ]
      }
      onSelect={[Function]}
      selectedInstruments={
        Array [
          "SHOC",
        ]
      }
      selectedTelescopes={
        Array [
          "Lesedi",
        ]
      }
    />
    <Filters
      filters={
        Array [
          "All",
        ]
      }
      instruments={
        Array [
          "SHOC",
        ]
      }
      onSelect={[Function]}
    />
  </styled.div>
</Fragment>
`;

exports[`Telescope form rendering should render correctly 3`] = `
<Fragment>
  <styled.div>
    <TelescopeSelector
      onSelect={[Function]}
      telescopes={
        Array [
          "SALT",
        ]
      }
    />
    <InstrumentSelector
      instruments={
        Array [
          "Salticam",
        ]
      }
      onSelect={[Function]}
      selectedTelescopes={
        Array [
          "SALT",
        ]
      }
    />
    <InstrumentModeSelector
      instrumentModes={
        Array [
          "All",
        ]
      }
      onSelect={[Function]}
      selectedInstruments={
        Array [
          "Salticam",
        ]
      }
      selectedTelescopes={
        Array [
          "SALT",
        ]
      }
    />
    <DetectorModeSelector
      detectorModes={
        Array [
          "All",
        ]
      }
      onSelect={[Function]}
      selectedInstruments={
        Array [
          "Salticam",
        ]
      }
      selectedTelescopes={
        Array [
          "SALT",
        ]
      }
    />
    <Filters
      filters={
        Array [
          "All",
        ]
      }
      instruments={
        Array [
          "Salticam",
        ]
      }
      onSelect={[Function]}
    />
  </styled.div>
</Fragment>
`;
